torch/distributed/_tensor/experimental/attention.py

```diff
@@ -361,153 +332,210 @@ def sdpa_backward_handler(
 
-            out_grad_queries.append(grad_query)
-            out_grad_keys.append(grad_key)
-            out_grad_values.append(grad_value)
+def _scaled_dot_product_ring_flash_attention_backward(
+    mesh: DeviceMesh,
+    grad_out: torch.Tensor,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    out: torch.Tensor,
+    logsumexp: torch.Tensor,
+    cum_seq_q: torch.Tensor,
+    cum_seq_k: torch.Tensor,
+    max_q: int,
+    max_k: int,
+    dropout_p: float,
+    is_causal: bool,
+    philox_seed: torch.Tensor,
+    philox_offset: torch.Tensor,
+    *,
+    scale: Optional[float] = None,
+) -> Tuple[torch.Tensor, ...]:
+    return _templated_ring_attention_backward(
+        mesh,
+        aten._scaled_dot_product_flash_attention_backward.default,
+        grad_out=grad_out,
+        grad_out_name="grad_out",
+        query=query,
+        key=key,
+        value=value,
+        out=out,
+        logsumexp=logsumexp,
+        is_causal=is_causal,
+        cum_seq_q=cum_seq_q,
+        cum_seq_k=cum_seq_k,
+        max_q=max_q,
+        max_k=max_k,
+        dropout_p=dropout_p,
+        philox_seed=philox_seed,
+        philox_offset=philox_offset,
+        scale=scale,
+    )
 
-    # stack and sum to avoid accumulation errors
-    out_grad_query = torch.stack(out_grad_queries).sum(dim=0)
-    out_grad_key = torch.stack(out_grad_keys).sum(dim=0)
-    out_grad_value = torch.stack(out_grad_values).sum(dim=0)
 
-    return out_grad_query, out_grad_key, out_grad_value
+def _scaled_dot_product_ring_efficient_attention_backward(
+    mesh: DeviceMesh,
+    grad_out: torch.Tensor,
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    bias: torch.Tensor,
+    out: torch.Tensor,
+    logsumexp: torch.Tensor,
+    philox_seed: torch.Tensor,
+    philox_offset: torch.Tensor,
+    dropout_p: float,
+    grad_input_mask: Tuple[bool, ...],
+    is_causal: bool = False,
+    *,
+    scale: Optional[float] = None,
+) -> Tuple[torch.Tensor, ...]:
+    return _templated_ring_attention_backward(
+        mesh,
+        aten._scaled_dot_product_efficient_attention_backward.default,
+        grad_out=grad_out,
+        grad_out_name="grad_out_",
+        query=query,
+        key=key,
+        value=value,
+        attn_bias=bias,
+        out=out,
+        logsumexp=logsumexp,
+        philox_seed=philox_seed,
+        philox_offset=philox_offset,
+        dropout_p=dropout_p,
+        grad_input_mask=grad_input_mask,
+        is_causal=is_causal,
+        scale=scale,
+    )
 
 
 customized_ops = {
     aten._scaled_dot_product_flash_attention.default: sdpa_handler,
     aten._scaled_dot_product_flash_attention_backward.default: sdpa_backward_handler,
+    aten._scaled_dot_product_efficient_attention.default: sdpa_handler,
+    aten._scaled_dot_product_efficient_attention_backward.default: sdpa_backward_handler,
 }
 
 
+# Following experimental APIs allow users to enable CP.
```

Comment:
looks like we have multiple user facing APIs: 1. attention_context_parallel 2. enable_context_parallel 3. context_parallel_buffers. It would be good to consolidate and remove some that are not necessary

Reply:
I merged and remove the APIs. Now we only have one public API, `context_parallel()`.